### PR TITLE
statistics: skip auto-analyze when it is executed outside the available time period

### LINKED ISSUE #49552: auto analyze will not stop outside the available timep period
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

create 500k tables.

set tidb_auto_analyze_start_time and tidb_auto_analyze_end_time

### 2. What did you expect to see? (Required)

it can be stopped.

### 3. What did you see instead (Required)

it cannot be stopped.

### PR DESCRIPTION
**Pull request by hawkingrei**
### What problem does this PR solve?

Issue Number: close #49552

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/statistics/handle/autoanalyze/autoanalyze.go [hunk 1]
```diff
@@ -392,6 +395,10 @@ func randomPickOneTableAndTryAutoAnalyze(
 
 		// We need to check every partition of every table to see if it needs to be analyzed.
 		for _, tbl := range tbls {
+			if !timeutil.WithinDayTimePeriod(start, end, time.Now()) {
+				statslogutil.StatsLogger().Info("auto analyze stopped", zap.String("reason", "not in the available time period"))
```
Comment (time-and-fate) [inline pkg/statistics/handle/autoanalyze/autoanalyze.go:399]: We are not printing logs in `HandleAutoAnalyze` when doing the same check.
I think printing a log here will be a bit confusing.

Comment (hawkingrei) [reply pkg/statistics/handle/autoanalyze/autoanalyze.go:399]: OK, I removed it.

File: pkg/statistics/handle/autoanalyze/autoanalyze.go [hunk 2]
```diff
@@ -335,28 +335,31 @@ func HandleAutoAnalyze(
...
 }
 
-// randomPickOneTableAndTryAutoAnalyze randomly picks one table and tries to analyze it.
+// RandomPickOneTableAndTryAutoAnalyze randomly picks one table and tries to analyze it.
 // 1. If the table is not analyzed, analyze it.
 // 2. If the table is analyzed, analyze it when "tbl.ModifyCount/tbl.Count > autoAnalyzeRatio".
 // 3. If the table is analyzed, analyze its indices when the index is not analyzed.
 // 4. If the table is locked, skip it.
-func randomPickOneTableAndTryAutoAnalyze(
+func RandomPickOneTableAndTryAutoAnalyze(
```
Comment (hi-rustin) [inline pkg/statistics/handle/autoanalyze/autoanalyze.go:355]:
```suggestion
// Exposed solely for testing.
func RandomPickOneTableAndTryAutoAnalyze(
```

File: pkg/statistics/handle/autoanalyze/autoanalyze.go [hunk 3]
```diff
@@ -392,6 +395,9 @@ func randomPickOneTableAndTryAutoAnalyze(
 
 		// We need to check every partition of every table to see if it needs to be analyzed.
 		for _, tbl := range tbls {
+			if !timeutil.WithinDayTimePeriod(start, end, time.Now()) {
```
Comment (hi-rustin) [inline pkg/statistics/handle/autoanalyze/autoanalyze.go:398]: Could you please comment on why it is necessary to check it repeatedly?

File: pkg/statistics/handle/autoanalyze/autoanalyze_test.go [hunk 4]
```diff
@@ -523,3 +525,21 @@ func TestCleanupCorruptedAnalyzeJobsOnDeadInstances(t *testing.T) {
...
+		dbName := fmt.Sprintf("db%d", i)
+		tk.MustExec(fmt.Sprintf("create database %s", dbName))
+		for j := 0; j < 2; j++ {
+			tableName := fmt.Sprintf("table%d", j)
+			tk.MustExec(fmt.Sprintf("create table %s.%s (a int)", dbName, tableName))
+		}
+	}
+	se, err := dom.SysSessionPool().Get()
+	require.NoError(t, err)
+	require.False(t, autoanalyze.RandomPickOneTableAndTryAutoAnalyze(se.(sessionctx.Context), dom.StatsHandle(), dom.SysProcTracker(), dom.InfoSchema(), 0.6, variable.Dynamic, ttStart, ttEnd))
```
Comment (hi-rustin) [inline pkg/statistics/handle/autoanalyze/autoanalyze_test.go:544]: This line appears to be a bit lengthy. Would it be better if we break it down?

**Review by hi-rustin — APPROVED** [review]
rest LGTM.

Approve in advance. Please fix the nit issues.

**Review by time-and-fate — APPROVED** [review]

File: pkg/statistics/handle/autoanalyze/autoanalyze_test.go (hunk 4 above)
Comment (hawkingrei) [reply pkg/statistics/handle/autoanalyze/autoanalyze_test.go]: fixed

File: pkg/statistics/handle/autoanalyze/autoanalyze.go (hunk 3 above)
Comment (hawkingrei) [reply pkg/statistics/handle/autoanalyze/autoanalyze.go:401]: OK, I added.

**Pull request merged**